### ISSUE TITLE
Standalone inline HTML tags rendered with extraneous DOM elements

### ISSUE THREAD

**Issue by justin-lau** [rexxars/react-markdown#43]
It appears to only happen to inline HTML tags (`<a>`, `<span>`, etc.) when they are put in a isolated lines.

### Example

This:

```markdown
<div>This is a normal line</div>

<a href="https://www.google.com">Google</a>
```

Gets compiled into this _(For the sake of readability I've removed all the `data-reactid`)_:

```javascript
<div>
    <div>This is a normal line</div>
</div>
<p>
    <span>
        <a href="https://www.google.com"></a>
    </span>
    <a href="https://www.google.com">
        <span>Google</span>
        <span></span>
    </a>
</p>
```

**Comment by rexxars** [comment]
Yes, this is a bit of a hard one to fix. Since markdown support HTML with markdown embedded inside of it (think `<a href="#">**hey bold**</a>`), all markdown parsers work in the way that they treat the opening and closing tags as separate tokens. So for the example above, it'll emit an `InlineHtml` block with the value of `<a href="#">`, then a `Emphasis` node with `hey bold` as value, then a new `InlineHtml` block with the value of `</a>`.

Unless we want to parse every HTML block and manually match opening and closing points, we're stuck with having to wrap the open and close tags in a span or similar, as is what has happened here.

This is one of my biggest annoyances with Markdown right now, but I have not felt up to the task of handling HTML parsing and the like. Super happy if anybody else wants to handle it, so leaving this open.

**Comment by alexlatchford** [comment]
Hi @rexxars just hitting this limitation now. Can you explain a bit more why the wrapping elements are needed at all? (ie. `<div>` or `<span>`). I think I'm missing the point of what can potentially happen if these are not enclosed.

Is it that the design decision you've made is to favour any potential inner markdown styling over the inline html? In @justin-lau's example there is no nested markdown so unsure why his example is affected.

**Comment by rexxars** [comment]
Unlike most react markdown libraries, this one doesn't take a string of markdown and parse it to a string of HTML, then put that HTML into your page with `innerHTML` or similar. Instead, it basically parses the markdown token by token and builds a corresponding tree of React components.

The benefit of this is that you will benefit from Reacts virtual DOM diffs, and each renderer can be customized just how you want them to work.

The drawback is that there isn't a clean way of parsing a block of raw HTML (say, `<a href="#">foo</a>`) to React nodes.  We could of course pull in a full HTML parser, parse the HTML and translate the AST to proper React nodes. This sort of works, but adds a large number of bytes to the bundle, makes the library much more complex, and will probably also pull in all sorts of requirements such as filtering the allowed nodes. There's also a question about what happens on invalid markup, say an unclosed tag.

**Comment by rgbkrk** [comment]
Inline tags can end up creating multiple nodes, which is the reason that attributes "aren't" being applied. This snippet in markdown:

```
<span style="color:red; font-family:Helvetica Neue, Helvetica, Arial, sans-serif; font-size:2em;">An Exception was encountered at 'In [16]'.</span>
```

Ends up hitting the html renderer twice, once with `<span style="color:red; font-family:Helvetica Neue, Helvetica, Arial, sans-serif; font-size:2em;">` followed by once with `</span>`. The text inside is a separate node, and we can't treat it as part of an html fragment.
